File: c_programming/common/n18_binary_number_with_alternating_bits_693.c
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include "utils.h"

static bool has_alternating_bits(int32_t n)
{
    bool ret = false;
    bool current_bit_status = false, last_bit_status = false;

    last_bit_status = n & 0x1;
    while (n) {
        n >>= 1;
        current_bit_status = n & 0x1;
        if (current_bit_status == last_bit_status) {
            return false;
        }
        last_bit_status = current_bit_status;
    }

    return true;
}

int32_t main(void)
{
    int32_t ret = 0;
    bool res = false;

    res = has_alternating_bits(0x55);
    LOG("the result is %d\n", res);

    UTILS_CHECK_RET(ret);

finish:
    return ret;
}

File: rust_programming/common/src/n18_binary_number_with_alternating_bits_693.rs

fn has_alternating_bits(n:i32) -> bool
{
    let mut last_bit_status:i32 = 0;
    let mut current_bit_status:i32 = 0;
    let mut e:i32 = n;

    last_bit_status = e & 0x1_i32;
    while e != 0 {
        e >>= 1;
        current_bit_status = e & 0x1_i32;
        if current_bit_status == last_bit_status {
            return false;
        }
        last_bit_status = current_bit_status;
    }

    return true;
}

pub fn test_has_alternating_bits()
{
    let res = has_alternating_bits(0x55);
    println!("the result is {res}");
}